tests/test_communication_manager.py

```diff
@@ -0,0 +1,163 @@
+
+def make_model_json():
+    m = KerasPerceptron(is_training=True)
+    model_architecture = m.model.to_json()
+    model_optimizer = get_optimizer(m.model)
+    model_json = {
+        "architecture": model_architecture,
+        "optimizer": model_optimizer
+    }
+    return model_json
+
+def make_initialize_job(model_json):
+    initialize_job = DMLJob(
+        "initialize",
+        model_json,
+        "keras",
+        hyperparams=make_hyperparams(),
+        label_column_name='label'
+    )
+    return initialize_job
+
+def make_hyperparams():
+    hyperparams = {
+        'averaging_type': 'data_size',
+        'batch_size': 4,
+        'epochs': 1,
+        'split': 0.004,
+    }
+    return hyperparams
+
+config_manager = ConfigurationManager()
+config_manager.bootstrap(
+    config_filepath='tests/artifacts/communication_manager/configuration.ini'
+)
+
+def test_communication_manager_can_initialize_and_train_model():
+    """
+    (1) CommMgr receives the packet it's going to receive from BlockchainGateway
+    (2) CommMgr gives packet to Optimizer
+    (3) Optimizer tells CommMgr to schedule an initialization job
+    (4) CommMgr schedules initialization job
+    (5) CommMgr receives DMLResult for initialization job
+    (6) CommMgr gives DMLResult to Optimizer
+    (7) Optimizer updates its weights to initialized model
+    (8) Optimizer tells CommMgr to schedule a training job
+    (9) CommMgr schedules training job
+    (10) CommMgr receives DMLResult for training job
+    (11) CommMgr gives DMLResult to Optimizer
+    (12) Optimizer updates its weights to trained weights
+    (13) Optimizer tells CommMgr to schedule a communication job
+    (14) CommMgr schedules communication job
+    (15) Exception raised, since communication job is not implemented
+    NOTE: No, I can't split this up. IDK why exactly but it hangs forever.
+    """
+    communication_manager = CommunicationManager(config_manager)
+    scheduler = DMLScheduler(config_manager, communication_manager)
+    scheduler.reset()
+    communication_manager.configure(scheduler)
+    new_session_event = {
+            "key": None, 
+            "content": {
+                "optimizer_params": {}, 
+                "serialized_job": serialize_job(make_initialize_job(make_model_json()))}}
+    communication_manager.inform(RawEventTypes.NEW_SESSION.value, new_session_event)
+    scheduler.start_cron(period_in_mins = 0.05)
+    time.sleep(10)
+    scheduler.stop_cron()
+    assert len(scheduler.processed) == 2
+    # TODO: I can't get the below AssertionError test to work since Georgy's code won't raise errors but it WORKS
+    # try:
+    #     scheduler.start_cron(period_in_mins = 0.05)
+    #     time.sleep(10)
+    #     scheduler.stop_cron()
+    #     assert len(scheduler.processed) == 2
+    #     assert False, "Runner should have been told to communicate"
+    # except AssertionError as e:
+    #     assert str(e) == "communication is not yet supported!"
```

Comment:
You're not going to be able to catch the exception, since the current functionality of the Runner catches errors and returns it in DMLResult. See line 131. If a job fails more than 3 times (see scheduler), then the job is discarded. 

So I'm not sure whether this test can be implemented as of now.

Reply:
sadboi